File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/oo/decorator/Examples.scala
package com.mblinn.mbfpp.oo.decorator

object Examples {

  def add(a: Int, b: Int) = a + b
  def subtract(a: Int, b: Int) = a - b
  def multiply(a: Int, b: Int) = a * b
  def divide(a: Int, b: Int) = a / b
  
  def makeLogger(calcFn: (Int, Int) => Int) = 
    (a: Int, b: Int) => {
      val result = calcFn(a, b)
      println("Result is: " + result)
      result
    }
  
  val loggingAdd = makeLogger(add)
  val loggingSubtract = makeLogger(subtract)
  val loggingMultiply = makeLogger(multiply)
  val loggingDivide = makeLogger(divide)
}

File: mbfpp/code/ClojureExamples/src/mbfpp/oo/decorator/examples.clj
(ns mbfpp.oo.decorator.examples)

(defn add [a b] (+ a b))
(defn subtract [a b] (- a b))
(defn multiply [a b] (* a b))
(defn divide [a b] (/ a b))

(defn make-logger [calc-fn]
  (fn [a b]
    (let [result (calc-fn a b)]
      (println (str "Result is: " result))
      result)))

(def logging-add (make-logger add))
(def logging-subtract (make-logger subtract))
(def logging-multiply (make-logger multiply))
(def logging-divide (make-logger divide))
